src/python/pants/backend/python/rules/resolve.py

```diff
@@ -0,0 +1,356 @@
+        ),
+        output_directories=(request.cache.path,),
+    )
+    resolve_result = await Get[ProcessResult](Process, execute_resolve)
+    dependencies = DistributionDependencies.load(StringIO(resolve_result.stdout.decode()))
+    cache_snapshot = await Get[Snapshot](
+        SnapshotSubset(
+            digest=resolve_result.output_digest,
+            globs=PathGlobs(
+                globs=[
+                    f"{request.cache.path}/**/*",
+                    # Since we do not attempt to leverage https://www.python.org/dev/peps/pep-0552/,
+                    # always discard bytecode cache files since they will contain a nondeterministic
+                    # timestamp.
+                    f"!{request.cache.path}/**/*.pyc",
+                    f"!{request.cache.path}/**/__pycache__",
+                ]
+            ),
+        )
+    )
+    return ResolveResult(dependencies=dependencies, cache=PexCache(snapshot=cache_snapshot))
+
+
+class ResolveOptions(Outputting, GoalSubsystem):
+    """Resolve python requirements."""
+
+    name = "resolve"
+
+    @classmethod
+    def register_options(cls, register) -> None:
+        super().register_options(register)
+        register(
+            "--local",
+            type=bool,
+            default=False,
+            help="Just resolve for a single local interpreter that matches compatibility "
+            "constraints, if any.",
+        )
+
+    @property
+    def local(self) -> bool:
+        return cast(bool, self.values.local)
+
+
+@rule
+async def resolve_target(options: ResolveOptions, target: PythonLibrary) -> ResolveResult:
+    # Recurse to get the resolution cache, if any.
+    direct_dependency_addresses = target[Dependencies].value
+    if not direct_dependency_addresses:
+        return ResolveResult()
+
+    requests: List[Get[Any]] = [
+        Get[Targets](Addresses(direct_dependency_addresses)),
+        Get[TransitiveTargets](Addresses([target.address])),
+    ]
+
+    (direct_dependencies, transitive_dependencies) = cast(
+        Tuple[Targets, TransitiveTargets], await MultiGet(requests)
+    )
+
+    requirements = tuple(
+        str(requirement.requirement)
+        for target in transitive_dependencies.closure
+        if isinstance(target, PythonRequirementLibrary)
+        for requirement in target[PythonRequirementsField].value
+    )
+    if not requirements:
+        return ResolveResult()
+
+    resolved_dependencies = await MultiGet(
+        Get[ResolveResult](PythonLibrary, dependency)
+        for dependency in direct_dependencies
+        if isinstance(dependency, PythonLibrary)
+    )
+
+    # The Pip http cache is non-deterministic. It records HTTP headers from responses which
+    # includes things like:
+    #
+    #   Date: Fri, 01 May 2020 17:20:41 GMT
+    #
+    # The non-determinism in these headers does not affect resolves, so we pick one winner when we
+    # find duplicates to ensure we can merge caches without conflict.
```

Comment:
Does this deserve an intrinsic or flag on merge directories (as a followup)? I'm pretty sure we don't want it to be the default behavior, but it seems like a reasonable case to support if someone opts in.